src/str.rs

```diff
@@ -0,0 +1,568 @@
+            Ok(ok) => ok,
+            Err(_) => unreachable!(),
+        }
+    }
+
+    /// Converts an octets builder into a string builder without checking.
+    ///
+    /// For the safe versions, see [from_utf8][Self::from_utf8],
+    /// [try_from_utf8_lossy][Self::try_from_utf8_lossy] and
+    /// [from_utf8_lossy][Self::from_utf8_lossy].
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `octets` contains data that is a correctly
+    /// UTF-8 encoded string. It may be empty.
+    pub unsafe fn from_utf8_unchecked(octets: Octets) -> Self {
+        Self(octets)
+    }
+
+    /// Converts the string builder into the underlying octets builder.
+    pub fn into_octets_builder(self) -> Octets {
+        self.0
+    }
+
+    /// Converts the string builder into the final str.
+    pub fn freeze(self) -> Str<Octets::Octets>
+    where Octets: OctetsBuilder {
+        Str(self.0.freeze())
+    }
+
+    /// Returns a slice of the already assembled string.
+    pub fn as_str(&self) -> &str
+    where Octets: AsRef<[u8]> {
+        unsafe { str::from_utf8_unchecked(self.0.as_ref()) }
+    }
+
+    /// Returns a mutable slice of the already assembled string.
+    pub fn as_str_mut(&mut self) -> &mut str
+    where Octets: AsMut<[u8]> {
+        unsafe { str::from_utf8_unchecked_mut(self.0.as_mut()) }
+    }
+
+    /// Returns the string’s octets as a slice.
+    pub fn as_slice(&self) -> &[u8]
+    where Octets: AsRef<[u8]> {
+        self.0.as_ref()
+    }
+
+    /// Returns the length of the string in octets.
+    pub fn len(&self) -> usize
+    where Octets: AsRef<[u8]> {
+        self.0.as_ref().len()
+    }
+
+    /// Returns whether the string is empty.
+    pub fn is_empty(&self) -> bool
+    where Octets: AsRef<[u8]> {
+        self.0.as_ref().is_empty()
+    }
+
+    /// Appends a given string slice onto the end of this builder.
+    pub fn try_push_str(
+        &mut self, s: &str,
+    ) -> Result<(), Octets::AppendError>
+    where Octets: OctetsBuilder {
+        self.0.try_append_slice(s.as_bytes())
+    }
+
+    /// Appends a given string slice onto the end of this builder.
+    pub fn push_str(&mut self, s: &str)
+    where
+        Octets: OctetsBuilder,
+        Octets::AppendError: Into<Infallible>,
+    {
+        self.0.append_slice(s.as_bytes())
+    }
+
+    /// Appends the given character to the end of the builder.
+    pub fn try_push(&mut self, ch: char) -> Result<(), Octets::AppendError>
+    where Octets: OctetsBuilder {
+        let mut buf = [0u8; 4];
+        self.0.try_append_slice(ch.encode_utf8(&mut buf).as_bytes())
+    }
+
+    /// Appends the given character to the end of the builder.
+    pub fn push(&mut self, ch: char)
+    where
+        Octets: OctetsBuilder,
+        Octets::AppendError: Into<Infallible>,
+    {
+        let mut buf = [0u8; 4];
+        self.0.append_slice(ch.encode_utf8(&mut buf).as_bytes())
+    }
+
+    /// Truncates the builder back to the given index.
```

Comment:
Possibly confusing, if I truncate back to 'index 2', can I store 2 or 3 elements?
Perhaps something like this is less ambiguous (inspired by https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate ):
```suggestion
    /// Truncates the builder, keeping the first `new_len` octets.
```

Reply:
Fair point. Fixed.